fix unregister_toolgroup error

pdettori opened the pull request:
# What does this PR do?
Fixes issue #1537 that causes "500 Internal Server Error" when unregistering a toolgroup

# (Closes #1537 )

## Test Plan

1. Start Ollama
2. On a Llama Stack development environment, run `llama stack run ollama`
3. On a different terminal, run a MCP server providing a tool. For this test, I have used  [mcp simple-tool](https://github.com/modelcontextprotocol/python-sdk/tree/main/examples/servers/simple-tool)
4. On a different terminal, run the following llama stack client SDK code to register and then unregister the tool:
```python
from llama_stack_client import LlamaStackClient
from llama_stack_client.types.shared_params.url import URL

client = LlamaStackClient(
    base_url="http://localhost:8321",
)

toolgroup_id = "remote::web-fetch"

client.toolgroups.register(
    toolgroup_id=toolgroup_id,
    provider_id="model-context-protocol",
    mcp_endpoint=URL(uri="http://localhost:8000/sse"),
)

client.toolgroups.unregister(toolgroup_id=toolgroup_id)
```
5. Verify that there are no errors with the fix. The same setup and code can be used that there is an error without the fix.

booxter reviewed (approved):
Indeed.

pdettori commented:
@ashwinb - test added in `tests/integration/tool_runtime/test_registration.py`

ashwinb commented on tests/integration/tool_runtime/test_registration.py:
```diff
@@ -0,0 +1,114 @@
+    )
+
+    def run_server():
+        uvicorn.run(app, host="0.0.0.0", port=8000)
```
I think this port should be a random port assignment in general (otherwise test could be flakey) -- but do we need to do all of this? Could we not use an appropriate fake / mock here -- the goal of this test is to check registration and unregistration.

pdettori replied:
@ashwinb I can make the port assigned randomly, as for mocking, is there some point in the code where we could hook up a mock HTTP server? Otherwise we would have a similar problem even if I just use a just a regular HTTP server, and I would have also to implement part of the MCP protocol in the mock. Also, I suppose that the MCP stdio protocol is not supported in the MCP provider, right?

leseb reviewed (commented):
Could you please rebase and then I'll trigger the tests? Thanks!

pdettori commented:
@leseb I have rebased and updated the tests to account for some changes in the API behavior (now `toolgroups.get(toolgroup_id=test_toolgroup_id)` throws an exception when `toolgroup_id` is not found). Also, I have added code to ensure an open port is found for the MCP test server, as requested by @ashwinb.

franciscojavierarceo commented on llama_stack/distribution/routers/routing_tables.py:
```diff
@@ -608,8 +608,8 @@ async def unregister_toolgroup(self, toolgroup_id: str) -> None:
-        tools = (await self.list_tools(toolgroup_id)).data
-        for tool in tools:
+        tools = await self.list_tools(toolgroup_id)
+        for tool in tools.data:
```
Would be good to handle case where data is not an attribute

pdettori replied:
@franciscojavierarceo - good point. Done.

franciscojavierarceo replied:
Thank you!

franciscojavierarceo reviewed (changes requested):
looks like there are some mypy failures, can you revise?

pdettori commented:
I see the issues on mypy - same issue running locally, and I did not touch the referenced file `llama_stack/templates/verification/verification.py` - something seems broke in main indeed.

```shell
llama_stack/templates/verification/verification.py:98: error: Incompatible return value type (got "tuple[list[Provider], dict[str, list[ProviderModelEntry]]]", expected "tuple[list[Provider], list[ModelInput]]")  [return-value]
llama_stack/templates/verification/verification.py:172: error: Argument 1 to "get_model_registry" has incompatible type "list[ModelInput]"; expected "dict[str, list[ProviderModelEntry]]"  [arg-type]
llama_stack/templates/verification/verification.py:180: error: Argument "available_models_by_provider" to "DistributionTemplate" has incompatible type "list[ModelInput]"; expected "dict[str, list[ProviderModelEntry]] | None"  [arg-type]
Found 3 errors in 1 file (checked 299 source files)
```

franciscojavierarceo reviewed (approved):
lgtm!

leseb reviewed (approved):
Thank you for your patience!
